File: demo/titanic_demo.py
#Author Benjamin Pachev <benjaminpachev@greenseedtechnologies.com>
#
#Python script to generate a simple XGBoost model for the titanic dataset

from pandas import read_csv
from numpy import array, mean
import xgboost as xgb

df = read_csv("titanic.csv")
labels = array(df["Survived"])
#Dummy code the Sex column
df["Sex_is_female"] = df["Sex"] == "female"

missing_val = -999
matrix = xgb.DMatrix(df[["Sex_is_female", "Age", "Fare"]].fillna(-999), label=labels, missing=missing_val)

params = {"max_depth":3, "num_rounds":10, "objective":"binary:logistic"}
model = xgb.train(params, matrix)

preds = model.predict(matrix)
pred_labels = preds > .5
print "Accuracy on train set {:.4f}%".format(100*mean(pred_labels==labels))
#Save a model for PHP to use
model.save_model("titanic.model")

File: demo/titanic_demo.php
<?php
define('TITANIC_NUM_FEATURE', 3);

function accuracy_score($labels, $preds)
{
    $total = count($preds);
    $num_correct = 0;
    foreach ($preds as $i => $pred)
    {
        $pred_label = ($pred > .5) ? 1 : 0;
        if ($pred_label == $labels[$i]) $num_correct += 1;
    }
    return $num_correct / $total;
}

function bail($msg)
{
    echo "$msg\n";
    exit(1);
}

function fix_null($v)
{
    #The default missing value for XGDMatrix is null, so we don't want to cast it to a float (which would make it zero).
    #XGBoost has special handling for missing values
    return (is_null($v)) ? $v : (float)$v;
}

function read_titanic_data(&$data, &$labels, $filename=__DIR__."/titanic.csv")
{
    if (!($file=fopen($filename, "r")))
    {
        bail("Error opening titanic dataset in $filename.");
    }
    #Read column names
    $column_names = fgetcsv($file);
    $names_to_inds = array_flip($column_names);

    $data = [];
    while (($row=fgetcsv($file)))
    {
        $is_female = ($row[$names_to_inds["Sex"]] == "female");
        $age = fix_null($row[$names_to_inds["Age"]]);
        $fare = fix_null($row[$names_to_inds["Fare"]]);
        $labels[] = (int)$row[$names_to_inds["Survived"]];
        $data[] = [$is_female, $age, $fare];
    }
}

read_titanic_data($data, $labels);

$matrix = new XGDMatrix($data, TITANIC_NUM_FEATURE);

$booster = new XGBooster();
$booster->loadModel(__DIR__."/titanic.model");
$preds = $booster->predict($matrix);

echo "Accuracy ".accuracy_score($labels, $preds)."\n";
$tree_limit = 1;
echo "Accuracy with one tree: ".accuracy_score($labels, $booster->predict($matrix, $tree_limit))."\n";
?>
